File: XTApp/XTCommon/XTUtility/XTLocationManger.m
//
//  XTLocationManger.m
//  XTApp
//
//  Created by xia on 2024/9/5.
//

#import "XTLocationManger.h"
#import <CoreLocation/CoreLocation.h>

@interface XTLocationManger()<CLLocationManagerDelegate>

@property (nonatomic, strong) CLLocationManager* manager;//用户位置服务

@end

@implementation XTLocationManger

+(instancetype)xt_share {
    static XTLocationManger *shareInstance;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        shareInstance = [[self alloc]init];
    });
    return shareInstance;
}

///是否可以使用位置服务
-(BOOL)xt_canLocation {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wundeclared-selector"
    //确定用户的位置服务启用
    BOOL can = [CLLocationManager locationServicesEnabled] && [CLLocationManager authorizationStatus]==kCLAuthorizationStatusDenied;
#pragma clang diagnostic pop
    if (can){
        return NO;
    }
    return YES;
}

///开始获取位置信息
- (BOOL)xt_startLocation {
    if(![self xt_canLocation]){
        return NO;
    }
    /** 由于IOS8中定位的授权机制改变 需要进行手动授权
        * 获取授权认证，两个方法：
        * [self.locationManager requestWhenInUseAuthorization];
        * [self.locationManager requestAlwaysAuthorization];
        */
   if ([self.manager respondsToSelector:@selector(requestAlwaysAuthorization)]) {
       XTLog(@"requestAlwaysAuthorization");
       [self.manager requestAlwaysAuthorization];
   }
   if ([self.manager respondsToSelector:@selector(requestWhenInUseAuthorization)]) {
       XTLog(@"requestAlwaysAuthorization");
       [self.manager requestWhenInUseAuthorization];
   }
   //开始定位，不断调用其代理方法
   [self.manager startUpdatingLocation];
    return YES;
}
- (void)locationManager:(CLLocationManager *)manager
     didUpdateLocations:(NSArray *)locations {
    //停止定位
    [self.manager stopUpdatingLocation];
    self.manager.delegate = nil;
    self.manager = nil;
    
    // 1.获取用户位置的对象
    CLLocation *location = [locations lastObject];
    CLLocationCoordinate2D coordinate = location.coordinate;
    self.xt_latitude = [NSString stringWithFormat:@"%f",coordinate.latitude];
    self.xt_longitude = [NSString stringWithFormat:@"%f",coordinate.longitude];
    if(self.LBSBlock) {
        self.LBSBlock();
    }
    if(self.LBSInfoBlock){
        CLGeocoder *geocoder=[[CLGeocoder alloc]init];
        @weakify(self)
        [geocoder reverseGeocodeLocation:location completionHandler:^(NSArray *placemarks,NSError *error) {
            @strongify(self)
             for(CLPlacemark *placemark in placemarks) {
                 //省
                 NSString *province = XT_Object_To_Stirng(placemark.administrativeArea);
                 //市
                 NSString *city = XT_Object_To_Stirng(placemark.locality);
                 NSDictionary *dic = @{
                     XTCountry:XT_Object_To_Stirng(placemark.country),
                     XTCountryCode:XT_Object_To_Stirng(placemark.ISOcountryCode),
                     XTProvince:province.length > 0 ? province : city,
                     XTCity:city.length > 0 ? city : province,
                     XTRegion:XT_Object_To_Stirng(placemark.subLocality),
                     XTStreet:XT_Object_To_Stirng(placemark.name),
                     XTLatitude:XT_Object_To_Stirng(self.xt_latitude),
                     XTLongitude:XT_Object_To_Stirng(self.xt_longitude),
                 };
                 if(self.LBSInfoBlock){
                     self.LBSInfoBlock(dic, YES);
                 }
                 return;
                 
             }
            if(self.LBSInfoBlock){
                self.LBSInfoBlock(@{}, NO);
            }
         }];
    }
    
    
}
- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error {
    if(self.LBSBlock){
        self.LBSBlock();
    }
    if(self.LBSInfoBlock){
        self.LBSInfoBlock(@{},NO);
    }
}

- (CLLocationManager *)manager{
    if(!_manager){
        _manager = [[CLLocationManager alloc] init];
        _manager.delegate = self;
        _manager.desiredAccuracy = kCLLocationAccuracyBest;
    }
    return _manager;
}

@end

File: XTApp/XTApp/XTCommon/XTUtility/XTLocationManger.swift
//
//  XTLocationManger.swift
//  XTApp
//
//  Created by Codex on 2026/4/26.
//

import CoreLocation
import Foundation

@objcMembers
@objc(XTLocationManger)
class XTLocationManager: NSObject, CLLocationManagerDelegate {
    static let shared = XTLocationManager()
    private var managerStorage: CLLocationManager?

    dynamic var xt_latitude: String?
    dynamic var xt_longitude: String?

    @objc(LBSBlock)
    dynamic var lbsBlock: XTBlock?

    @objc(LBSInfoBlock)
    dynamic var lbsInfoBlock: (([AnyHashable: Any], Bool) -> Void)?

    @objc class func xt_share() -> XTLocationManager {
        shared
    }

    @objc func xt_canLocation() -> Bool {
        let denied = CLLocationManager.locationServicesEnabled() && CLLocationManager.authorizationStatus() == .denied
        return !denied
    }

    @objc func xt_startLocation() -> Bool {
        guard xt_canLocation() else {
            return false
        }
        if manager.responds(to: #selector(CLLocationManager.requestAlwaysAuthorization)) {
            NSLog("requestAlwaysAuthorization")
            manager.requestAlwaysAuthorization()
        }
        if manager.responds(to: #selector(CLLocationManager.requestWhenInUseAuthorization)) {
            NSLog("requestAlwaysAuthorization")
            manager.requestWhenInUseAuthorization()
        }
        manager.startUpdatingLocation()
        return true
    }

    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        managerStorage?.stopUpdatingLocation()
        managerStorage?.delegate = nil
        managerStorage = nil

        guard let location = locations.last else {
            lbsBlock?()
            lbsInfoBlock?([:], false)
            return
        }

        let coordinate = location.coordinate
        xt_latitude = String(format: "%f", coordinate.latitude)
        xt_longitude = String(format: "%f", coordinate.longitude)
        lbsBlock?()

        guard lbsInfoBlock != nil else {
            return
        }

        let geocoder = CLGeocoder()
        geocoder.reverseGeocodeLocation(location) { [weak self] placemarks, _ in
            guard let self else { return }
            for placemark in placemarks ?? [] {
                let province = self.xtLocationString(placemark.administrativeArea)
                let city = self.xtLocationString(placemark.locality)
                let dic: [AnyHashable: Any] = [
                    "XTCountry": self.xtLocationString(placemark.country),
                    "XTCountryCode": self.xtLocationString(placemark.isoCountryCode),
                    "XTProvince": province.isEmpty ? city : province,
                    "XTCity": city.isEmpty ? province : city,
                    "XTRegion": self.xtLocationString(placemark.subLocality),
                    "XTStreet": self.xtLocationString(placemark.name),
                    "XTLatitude": self.xtLocationString(self.xt_latitude),
                    "XTLongitude": self.xtLocationString(self.xt_longitude)
                ]
                self.lbsInfoBlock?(dic, true)
                return
            }
            self.lbsInfoBlock?([:], false)
        }
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        lbsBlock?()
        lbsInfoBlock?([:], false)
    }

    private var manager: CLLocationManager {
        if managerStorage == nil {
            let locationManager = CLLocationManager()
            locationManager.delegate = self
            locationManager.desiredAccuracy = kCLLocationAccuracyBest
            managerStorage = locationManager
        }
        return managerStorage!
    }

    private func xtLocationString(_ value: Any?) -> String {
        guard let value, !(value is NSNull) else { return "" }
        if let string = value as? String {
            return string
        }
        return "\(value)"
    }
}
